src/main/java/pl/javastart/streamsexercise/PaymentService.java

```diff
@@ -19,70 +22,115 @@ class PaymentService {
+        return x -> x.getPaymentDate().getYear() == yearMonth.getYear()
+                && x.getPaymentDate().getMonth().ordinal() == yearMonth.getMonth().ordinal();
     }
 
     /*
     Znajdź i zwróć płatności dla ostatnich X dzni
      */
     List<Payment> findPaymentsForGivenLastDays(int days) {
-        throw new RuntimeException("Not implemented");
+        Stream<Payment> paymentStream = paymentRepository.findAll().stream();
+        ZonedDateTime zonedDateTimeNow = dateTimeProvider.zonedDateTimeNow();
+        ZonedDateTime zonedDateTimeMinusDaysFromUser = zonedDateTimeNow.minusDays(days);
+        return paymentStream
+                .filter(x -> x.getPaymentDate().isAfter(zonedDateTimeMinusDaysFromUser))
+                .toList();
     }
 
     /*
     Znajdź i zwróć płatności z jednym elementem
      */
     Set<Payment> findPaymentsWithOnePaymentItem() {
-        throw new RuntimeException("Not implemented");
+        return paymentRepository.findAll().stream()
+                .filter(x -> x.getPaymentItems().size() == 1).collect(Collectors.toSet());
     }
 
     /*
     Znajdź i zwróć nazwy produktów sprzedanych w aktualnym miesiącu
      */
     Set<String> findProductsSoldInCurrentMonth() {
-        throw new RuntimeException("Not implemented");
+        Stream<Payment> paymentStream = findPaymentsForCurrentMonth().stream();
+        Stream<PaymentItem> paymentItemStream = paymentStream.map(Payment::getPaymentItems)
+                .flatMap(Collection::stream);
+        return paymentItemStream.map(PaymentItem::getName).collect(Collectors.toSet());
     }
 
     /*
     Policz i zwróć sumę sprzedaży dla wskazanego miesiąca
      */
     BigDecimal sumTotalForGivenMonth(YearMonth yearMonth) {
-        throw new RuntimeException("Not implemented");
+        Stream<Payment> paymentStream = findPaymentsForGivenMonth(yearMonth).stream();
+        Stream<PaymentItem> paymentItemStream = paymentStream
+                .map(Payment::getPaymentItems)
+                .flatMap(Collection::stream);
+        Stream<BigDecimal> bigDecimalStream = paymentItemStream.map(PaymentItem::getFinalPrice);
+        return bigDecimalStream
+                .reduce(BigDecimal::add)
+                .get();
     }
 
     /*
     Policz i zwróć sumę przeyznanaych rabatów dla wskazanego miesiąca
      */
     BigDecimal sumDiscountForGivenMonth(YearMonth yearMonth) {
-        throw new RuntimeException("Not implemented");
+        Stream<Payment> paymentStream1 = findPaymentsForGivenMonth(yearMonth).stream();
+        Stream<PaymentItem> paymentItemStream = paymentStream1
+                .map(Payment::getPaymentItems)
+                .flatMap(Collection::stream);
+        Stream<BigDecimal> bigDecimalStream = paymentItemStream.map(x ->
+                x.getRegularPrice().subtract(x.getFinalPrice()));
+        return bigDecimalStream
+                .reduce(BigDecimal::add)
+                .get();
     }
```

Comment:
Tutaj ponownie, nie jest to błąd, ale przy strumieniach raczej wywołujemy kolejne metody na strumieniu, dokonując jego transformacji, bez dodatkowych zmiennych po wykonaniu każdej operacji